### PR TITLE
feat: Support returning a txHash asap, show a notification for a cancelled smart transaction

### PR DESCRIPTION
**Pull request by dan437**
## **Description**
This PR does a few things to support returning a txHash asap after a smart transaction is submitted:
- Shows a notification for a cancelled smart transaction
- Requires to send these params when calling the Smart Transactions controller: `clientId`, `getFeatureFlags` and `updateTransaction`

## **Related issues**

Resolves this JIRA ticket: https://consensyssoftware.atlassian.net/browse/TXL-517

## **Manual testing steps**

1. A feature flag for returning a txHash asap has to be set to true.
2. It can be tested e.g. via a dapp transaction. A dapp should receive a txHash asap right after submitting a smart transaction 
3. We show a Toast notification for cancelled smart transaction (e.g. when very low gas fees are used) if a user closed the Smart Transaction status page 

## **Screenshots/Recordings**
Success Toast:
image

Pending smart transaction:
image

Cancelled smart transaction notification:
image

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Review by klejeune — DISMISSED** [review]

**Review by antonydenyer — APPROVED** [review]

File: app/core/NotificationsManager.test.ts [hunk 1]
```diff
@@ -123,4 +126,186 @@ describe('NotificationManager', () => {
...
+        removeNotificationById,
+      });
+
+      // Create spy for _showNotification
+      showNotificationSpy = jest.spyOn(notificationManager, '_showNotification');
+    });
+
+    afterEach(() => {
+      showNotificationSpy.mockRestore();
+    });
```
Comment (NicolasMassart) [inline app/core/NotificationsManager.test.ts:197]: Q: why creating mock on before and then restore (like reset) after each? Why not just create them once and simply clean mocks to remove interactions but without create and destroy?

File: app/core/NotificationManager.js [hunk 2]
```diff
@@ -389,6 +395,37 @@ class NotificationManager {
...
+    const smartTransactionListener = async (smartTransaction) => {
+      if (smartTransaction.status === SmartTransactionStatuses.PENDING) {
+        return;
+      }
+      Engine.controllerMessenger.unsubscribe(
+        'SmartTransactionsController:smartTransaction',
+        smartTransactionListener,
+      );
+      if (smartTransaction.status !== SmartTransactionStatuses.CANCELLED) {
+        // If the smart transaction is not cancelled, notifications are already handled.
```
Comment (NicolasMassart) [inline app/core/NotificationManager.js:408]: Q: not sure to understand this comment. "already handled" where and why?

File: app/util/smart-transactions/smart-publish-hook.test.ts
Comment (NicolasMassart) [inline app/util/smart-transactions/smart-publish-hook.test.ts:1]: Q: Were do you test the new "asap" aspect of this depending on the feature flag?

File: app/core/NotificationManager.js (hunk 2 above)
Comment (dan437) [reply app/core/NotificationManager.js:408]: Already handled by the TransactionController by events like `TransactionController:transactionConfirmed`., etc. However, this use case of a cancelled smart transaction wasn't handled, that's why a listener for a SmartTransactionsController event was added.

File: app/util/smart-transactions/smart-publish-hook.test.ts
Comment (dan437) [reply app/util/smart-transactions/smart-publish-hook.test.ts:1]: It's possible to enforce it for testing purposes in the `/app/reducers/swaps/index.js` file in this code:
```
export const selectSwapsChainFeatureFlags = createSelector(
  swapsStateSelector,
  chainIdSelector,
  (swapsState, chainId) => ({
    ...swapsState[chainId].featureFlags,
    smartTransactions: {
      ...(swapsState[chainId].featureFlags?.smartTransactions || {}),
      ...(swapsState.featureFlags?.smartTransactions || {}),
      mobileReturnTxHashAsap: true,
    },
  }),
);
```

**Review by klejeune — APPROVED** [review]

**Review by wachunei — APPROVED** [review]

File: package.json [hunk 3]
```diff
@@ -182,7 +182,7 @@
     "@metamask/selected-network-controller": "^18.0.2",
     "@metamask/signature-controller": "^22.0.0",
     "@metamask/slip44": "3.1.0",
-    "@metamask/smart-transactions-controller": "^14.0.0",
+    "@metamask/smart-transactions-controller": "^15.0.0",
```
Comment (tommasini) [inline package.json:185]: Mobile have Network Controller at version 21, it seems that smart transactions controller at version 15 requires it to be at version 22
Also it expects the Transaction Controller to be at version 38 and it's currently at version 39
https://github.com/MetaMask/smart-transactions-controller/blob/main/CHANGELOG.md

Isn't that a blocker?

File: app/core/Engine/Engine.ts [hunk 4]
```diff
@@ -163,7 +163,7 @@ import SmartTransactionsController from '@metamask/smart-transactions-controller
 import { getAllowedSmartTransactionsChainIds } from '../../../app/constants/smartTransactions';
 import { selectShouldUseSmartTransaction } from '../../selectors/smartTransactionsController';
 import { selectSwapsChainFeatureFlags } from '../../reducers/swaps';
-import { SmartTransactionStatuses } from '@metamask/smart-transactions-controller/dist/types';
+import { SmartTransactionStatuses, ClientId } from '@metamask/smart-transactions-controller/dist/types';
```
Comment (tommasini) [inline app/core/Engine/Engine.ts:166]: Is there a plan to stop importing this from the dist and importing directly from the package?

**Comment by tommasini** [comment]
Would be very nice to have some recordings in the PR description

File: package.json (hunk 3 above)
Comment (Gudahtt) [reply package.json:185]: Note that it was already out of sync here; version 14 of the smart transactions controller already depended on network controller v21.1.0 which is a deprecated release equivalent to v22.

File: app/core/Engine/Engine.ts (hunk 4 above)
Comment (dan437) [reply app/core/Engine/Engine.ts:166]: Great idea, will create a ticket for it and we will do it in another PR.

**Review by tommasini — APPROVED** [review]

**Pull request merged**